File: MiniDOMBase/DMD/Anagrafica/CUserUffici.vb
Imports System.Net.Mail 'importo il Namespace
Imports System.Net.Sockets
Imports minidom
Imports minidom.Databases
Imports minidom.Anagrafica
Imports minidom.Sistema

Partial Public Class Anagrafica

    ''' <summary>
    ''' Collezione degli uffici a cui appartiene un utente
    ''' </summary>
    ''' <remarks></remarks>
    <Serializable> _
    Public Class CUserUffici
        Inherits CCollection(Of CUfficio)

        <NonSerialized> Private m_User As CUser

        Public Sub New()
            Me.m_User = Nothing
        End Sub

        Public Sub New(ByVal user As CUser)
            Me.New()
            Me.Load(user)
        End Sub

        Public ReadOnly Property User As CUser
            Get
                Return Me.m_User
            End Get
        End Property

        Private Shadows Sub Insert(ByVal index As Integer, ByVal u As CUfficio)
            SyncLock Me
                MyBase.Insert(index, u)
            End SyncLock
        End Sub

        Public Shadows Sub Add(ByVal u As CUfficio)
            If (u Is Nothing) Then Throw New ArgumentNullException("ufficio")
            Dim db As CDBConnection = Anagrafica.Uffici.GetConnection
            If (db.IsRemote) Then
                Dim uID As Integer = GetID(u)
                If (Me.GetItemById(uID) IsNot Nothing) Then Throw New Exception("La collezione contiene già l'ufficio specificato")
                Dim tmp As String = db.InvokeMethod(Anagrafica.Uffici.Module, "AddUfficio", "uid", RPC.int2n(GetID(Me.m_User)), "oid", uID)
                If (tmp <> "") Then Throw New Exception(tmp)
            Else
                SyncLock Me.m_User
                    Dim i As Integer = Me.IndexOf(u)
                    If (i >= 0) Then Throw New ArgumentException("L'ufficio è già associato all'utente")
                    SyncLock Anagrafica.Uffici
                        Dim item As CUtenteXUfficio = Nothing
                        For Each uxu As CUtenteXUfficio In Anagrafica.Uffici.UfficiConsentiti
                            If (uxu.IDUfficio = GetID(u)) AndAlso (uxu.IDUtente = GetID(Me.m_User)) Then
                                item = uxu
                                Exit For
                            End If
                        Next
                        If (item Is Nothing) Then
                            item = New CUtenteXUfficio
                            item.Ufficio = u
                            item.Utente = Me.m_User
                            item.Save()
                            Anagrafica.Uffici.UfficiConsentiti.Add(item)
                        End If
                        MyBase.Add(u)
                    End SyncLock
                End SyncLock
            End If
        End Sub

        Public Shadows Sub Remove(ByVal u As CUfficio)
            SyncLock Me.m_User
                Dim i As Integer = Me.IndexOf(u)
                If (i < 0) Then Throw New ArgumentException("L'ufficio non è associato all'utente")
                SyncLock Anagrafica.Uffici
                    Dim item As CUtenteXUfficio = Nothing
                    For Each uxu In Anagrafica.Uffici.UfficiConsentiti
                        If (uxu.IDUfficio = GetID(u)) AndAlso (uxu.IDUtente = GetID(Me.m_User)) Then
                            item = uxu
                            Exit For
                        End If
                    Next
                    If (item IsNot Nothing) Then
                        item.Delete()
                        Anagrafica.Uffici.UfficiConsentiti.Remove(item)
                    End If
                End SyncLock
                Me.RemoveAt(i)
            End SyncLock
        End Sub

        Public Shadows Function IndexOf(ByVal u As CUfficio) As Integer
            SyncLock Me.m_User
                If (u Is Nothing) Then Throw New ArgumentNullException("u")
                For i As Integer = 0 To Me.Count - 1
                    If GetID(Me(i)) = GetID(u) Then Return i
                Next
                Return -1
            End SyncLock
        End Function

        Protected Friend Function Load(ByVal user As CUser) As Boolean
            If (user Is Nothing) Then Throw New ArgumentNullException("user")
            SyncLock user
                SyncLock Anagrafica.Uffici
                    MyBase.Clear()
                    Me.m_User = user
                    If (GetID(user) = 0) Then Return True
                    For i As Integer = 0 To Anagrafica.Uffici.UfficiConsentiti.Count - 1
                        Dim item As CUtenteXUfficio = Anagrafica.Uffici.UfficiConsentiti(i)
                        If (item.IDUtente = GetID(user)) AndAlso (item.Ufficio IsNot Nothing) AndAlso (item.Ufficio.Stato = ObjectStatus.OBJECT_VALID) Then
                            MyBase.Add(item.Ufficio)
                        End If
                    Next
                    Return True
                End SyncLock
            End SyncLock
        End Function

        ''' <summary>
        ''' Restituisce vero se l'utente corrente condivide almeno un ufficio con l'utente specificato
        ''' </summary>
        ''' <param name="user"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Function SameOffice(ByVal user As CUser) As Boolean
            If (user Is Me.m_User) Then Return True
            If (user Is Nothing) Then Throw New ArgumentNullException("user")
            SyncLock Me.m_User
                SyncLock user
                    For i As Integer = 0 To Me.Count - 1
                        Dim u As CUfficio = Me(i)
                        For j As Integer = 0 To user.Uffici.Count - 1
                            Dim u1 As CUfficio = user.Uffici(j)
                            If GetID(u) = GetID(u1) Then Return True
                        Next
                    Next
                    Return False
                End SyncLock
            End SyncLock
        End Function

        Public Function HasOffice(ByVal office As CUfficio) As Boolean
            Return Me.HasOffice(GetID(office))
        End Function

        Public Function HasOffice(ByVal officeID As Integer) As Boolean
            SyncLock Me.m_User
                For i As Integer = 0 To Me.Count - 1
                    Dim u As CUfficio = Me(i)
                    If GetID(u) = officeID Then Return True
                Next
                Return False
            End SyncLock
        End Function

        Public Shadows Function Contains(ByVal u As CUfficio) As Boolean
            Return Me.IndexOf(u) >= 0
        End Function

        Protected Friend Overridable Sub SetUser(ByVal user As CUser)
            Me.m_User = user
        End Sub



    End Class

End Class

File: MiniDOMBase/DMD/Sistema/Users/Uffici/CUserUffici.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.repositories;

namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Collezione degli uffici a cui appartiene un utente
        /// </summary>
        /// <remarks></remarks>
        [Serializable]
        public class CUserUffici 
            : CCollection<CUfficio>
        {
            [NonSerialized] private Sistema.CUser m_User;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CUserUffici()
            {
                m_User = null;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="user"></param>
            public CUserUffici(Sistema.CUser user) : this()
            {
                Load(user);
            }

            /// <summary>
            /// Restituisce un riferimento all'utente
            /// </summary>
            public Sistema.CUser User
            {
                get
                {
                    return m_User;
                }
            }

            /// <summary>
            /// OnSet
            /// </summary>
            /// <param name="index"></param>
            /// <param name="oldValue"></param>
            /// <param name="newValue"></param>
            protected override void OnSet(int index, CUfficio oldValue, CUfficio newValue)
            {
                //if (this.m_User is object)
                //    newValue.SetUser(this.m_User);
                base.OnSet(index, oldValue, newValue);
            }

            /// <summary>
            /// OnInsert
            /// </summary>
            /// <param name="index"></param>
            /// <param name="value"></param>
            protected override void OnInsert(int index, CUfficio value)
            {
                //if (value is object)
                //    value.SetUser(this.m_User);
                base.OnInsert(index, value);
            }

            /// <summary>
            /// Inserisce l'ufficio
            /// </summary>
            /// <param name="index"></param>
            /// <param name="u"></param>
            private new void Insert(int index, CUfficio u)
            {
                lock (this)
                    base.Insert(index, u);
            }

            /// <summary>
            /// Aggiunge l'ufficio
            /// </summary>
            /// <param name="u"></param>
            public new void Add(CUfficio u)
            {
                if (u is null)
                    throw new ArgumentNullException("ufficio");
                //var db = Uffici.GetConnection();
                //if (db.IsRemote())
                //{
                //    int uID = DBUtils.GetID(u);
                //    if (GetItemById(uID) is object)
                //        throw new Exception("La collezione contiene già l'ufficio specificato");
                //    string tmp = db.InvokeMethod(Uffici.Module, "AddUfficio", "uid", Sistema.RPC.int2n(DBUtils.GetID(m_User)), "oid", uID);
                //    if (!string.IsNullOrEmpty(tmp))
                //        throw new Exception(tmp);
                //}
                //else
                //{
                    lock (m_User)
                    {
                        int i = IndexOf(u);
                        if (i >= 0)
                            throw new ArgumentException("L'ufficio è già associato all'utente");
                        lock (Uffici)
                        {
                            CUtenteXUfficio item = null;
                            foreach (var uxu in minidom.Anagrafica.Uffici.UfficiConsentiti)
                            {
                                if (
                                       uxu.IDUfficio == DBUtils.GetID(u, 0) 
                                    && uxu.IDUtente == DBUtils.GetID(m_User, 0)
                                    )
                                {
                                    item = uxu;
                                    break;
                                }
                            }

                            if (item is null)
                            {
                                item = new CUtenteXUfficio();
                                item.Ufficio = u;
                                item.Utente = m_User;
                                item.Save();
                                //Uffici.UfficiConsentiti.Add(item);
                            }

                            base.Add(u);
                        }
                    }
                //}
            }

            /// <summary>
            /// Rimuove l'ufficio
            /// </summary>
            /// <param name="u"></param>
            public new void Remove(CUfficio u)
            {
                lock (m_User)
                {
                    int i = IndexOf(u);
                    if (i < 0)
                        throw new ArgumentException("L'ufficio non è associato all'utente");
                    lock (Uffici)
                    {
                        CUtenteXUfficio item = null;
                        foreach (var uxu in Uffici.UfficiConsentiti)
                        {
                            if (uxu.IDUfficio == DBUtils.GetID(u, 0) && uxu.IDUtente == DBUtils.GetID(m_User, 0))
                            {
                                item = uxu;
                                break;
                            }
                        }

                        if (item is object)
                        {
                            item.Delete();
                            //Uffici.UfficiConsentiti.Remove(item);
                        }
                    }

                    RemoveAt(i);
                }
            }

            /// <summary>
            /// Restituisce l'indice dell'ufficio
            /// </summary>
            /// <param name="u"></param>
            /// <returns></returns>
            public new int IndexOf(CUfficio u)
            {
                lock (m_User)
                {
                    if (u is null)
                        throw new ArgumentNullException("u");
                    for (int i = 0, loopTo = Count - 1; i <= loopTo; i++)
                    {
                        if (DBUtils.GetID(this[i]) == DBUtils.GetID(u))
                            return i;
                    }

                    return -1;
                }
            }

            /// <summary>
            /// Carica gli uffici
            /// </summary>
            /// <param name="user"></param>
            /// <returns></returns>
            protected internal bool Load(Sistema.CUser user)
            {
                if (user is null)
                    throw new ArgumentNullException("user");
                lock (user)
                {
                    lock (Uffici)
                    {
                        Clear();
                        m_User = user;
                        if (DBUtils.GetID(user, 0) == 0)
                            return true;
                        
                        foreach(var item in  minidom.Anagrafica.Uffici.UfficiConsentiti)
                        { 
                            if (  
                                   item.IDUtente == DBUtils.GetID(user, 0) 
                                && item.Ufficio is object 
                                && item.Ufficio.Stato == ObjectStatus.OBJECT_VALID
                                )
                            {
                                base.Add(item.Ufficio);
                            }
                        }

                        return true;
                    }
                }
            }

            /// <summary>
            /// Restituisce vero se l'utente corrente condivide almeno un ufficio con l'utente specificato
            /// </summary>
            /// <param name="user"></param>
            /// <returns></returns>
            /// <remarks></remarks>
            public bool SameOffice(Sistema.CUser user)
            {
                if (ReferenceEquals(user, m_User))
                    return true;
                if (user is null)
                    throw new ArgumentNullException("user");
                lock (m_User)
                {
                    lock (user)
                    {
                        for (int i = 0, loopTo = Count - 1; i <= loopTo; i++)
                        {
                            var u = this[i];
                            for (int j = 0, loopTo1 = user.Uffici.Count - 1; j <= loopTo1; j++)
                            {
                                var u1 = user.Uffici[j];
                                if (DBUtils.GetID(u) == DBUtils.GetID(u1))
                                    return true;
                            }
                        }

                        return false;
                    }
                }
            }

            /// <summary>
            /// Restituisce true se l'ufficio appartiene alla collezione
            /// </summary>
            /// <param name="office"></param>
            /// <returns></returns>
            public bool HasOffice(CUfficio office)
            {
                return HasOffice(DBUtils.GetID(office, 0));
            }

            /// <summary>
            /// Restituisce true se l'ufficio appartiene alla collezione
            /// </summary>
            /// <param name="officeID"></param>
            /// <returns></returns>
            public bool HasOffice(int officeID)
            {
                lock (m_User)
                {
                    for (int i = 0, loopTo = Count - 1; i <= loopTo; i++)
                    {
                        var u = this[i];
                        if (DBUtils.GetID(u, 0) == officeID)
                            return true;
                    }

                    return false;
                }
            }

            /// <summary>
            /// Restituisce true se l'ufficio appartiene alla collezione
            /// </summary>
            /// <param name="u"></param>
            /// <returns></returns>
            public new bool Contains(CUfficio u)
            {
                return IndexOf(u) >= 0;
            }

            /// <summary>
            /// Imposta l'utente
            /// </summary>
            /// <param name="user"></param>
            protected internal virtual void SetUser(Sistema.CUser user)
            {
                m_User = user;
            }
        }
    }
}